### ISSUE TITLE
Restructuring of modules

### ISSUE THREAD

**Issue by chille** [olofson/eel#118]
I think it is time to take a look at the module system.

In my opinion the core should be really slim and have a low memory footprint. It should only contain the things really required by the EEL laungage. It should be possible to run it on a microcontroller without an operating system or filesystem and with limited resources in terms of storage and RAM. I'm thinking about a more powerful ARM Cortex M3 or similar.

To make that possible, we need to think about what modules should be a part of the core. All "extra" functionallity should be loaded as a module. I don't think we need dynamic loading of modules (yet), but it should be easy to choose what functionallity you include when integrating EEL into your application.

Module  | Comment
--------|-----------
builtin | Should platform specific function like ShellExecute really be here? Maybe it fits better in the system module?
dsp     | Move to an external module
dir     | Move to an external module
io      | Move to an external module
math    | Move to an external module. Mostly because I think this kind of mathematics have a high memory footprint when running on a microcontroller with limited resources.
system  | The only thing really useful seems to be the ARCH constant, which could be defined in e_state.c instead. Everything else could be moved into an external module
eelium  | Already an external module, but could probably be splitted into a few separate modules

All modules above should be included in src/executive/eel.c by default
Only the "builtin" module should be included in eel_open()

The eelium module consists of both C and EEL. I'm not really sure about the module concept. Could a "module" be just plain EEL, or does it need to have have mappings to C function to be a module?

**Comment by olofson** [comment]
Actually, I don't really want *any* modules to be hardwired parts of the core. Even the module load/import system could be implemented via the host application (much like how the Eelium and other modules are "injected" now), so that EEL could work without file systems, dynamic library loading etc.

The main reason they're all more or less integrated now is that I ran into various issues when implementing dynamic binary module loading the first time around, and as I didn't strictly need it at the time, I dropped it. One of those issues - SDL 1.2 initialization - will be gone as I finish up #71. Some other issues were eliminated in the module load/import subsystem refactoring a good while ago.

One remaining problem is that shared libraries linking to symbols exported by the loading executable is not portable. (This happens when a binary EEL module tries to call EEL core functions when EEL is statically linked into the host application. It *should* be fine as long as EEL is a DLL/so/dylib...) The most portable solution is probably to provide the EEL core API via a struct of function pointers or similar, passed to the module init function.

As for modules, in short, everything you load is a module, and they can export native functions, EEL functions, or both. I suspect some more refactoring, renaming, and/or clarification might be in place, but the general idea is that C and EEL modules and functions should be equivalent as far as possible, so that EEL code can transparently be replaced with a native code, and vice versa.

No objections regarding the respective modules.

ShellExecute() should definitely not be in builtin! print() and all that probably shouldn't be there either (or should have an extra level of abstraction), as that just causes unnecessary and platform specific issues when a host application wants to redirect I/O somewhere. Much of this was "designed" and implemented in a bit of a hurry, as I had more pressing problems to solve at the time.

As for Eelium, the whole thing is a bit of a mess, as it was mostly cobbled together for the lab instrument firmware + graphical calibration tools that I worked on back when I did most of the work on EEL, and later, physics, OpenGL and whatnot was added for Kobo II.

**Comment by chille** [comment]
I have created a new branch and started working with this:
https://github.com/chille/eel/tree/module-loader

What I have done so far
 * Compiler is removed from the core and put in a separate library libeelcompiler (There is still dependencies between them that needs to be sorted out)
 * Core is renamed from libeel to libeelcore
 * All modules are put in the ./src/modules directory and compiled into separate libraries, libeelmodulexxx
 * The calls for eel_*_init() does use printf() instead of eel_msg() for error handling (I'm not sure if best practice is to still use eel_msg()?)
 * system() and ShellExecute() are moved to the "system" module, instead of "builtin"
 * All modules are included in "eeltest" and "executive"
 * Most of the module loading code is moved to a new module named "loader"
 * EEL_state have a new "EEL_object *loader", which works the same way as the previously used "eelib" (Which is still there for __cleanup())

What still needs to be done
 * Eelium should be splitted into a few modules and moved into the ./src/modules/ directory, but I do not want to interfere with the work you are doing right now.
 * This code is leaking memory, I get an "VM WARNING: Classes CCLASS and CSTRING refcounts are 6 and 205 respectively" when running eeltest
 * Test everything on Windows
 * I have only compiled with option(BUILD_EELIUM ... OFF)
 * Check that "make install" works
 * The problem with static/dynamic linking you talking about above
 * Some code review wouldn't hurt...

I agree that print() should have some level of abstraction. I think this should be splitted into two different issues:
 * An abstraction layer for all code in core/compiler that require some function provided by an specific operating system and/or hardware. This would remove all #ifdef's for _WIN32_, __MACOSX__ and similar.
 * An abstraction layer for printing text. Both from inside EEL scripts and from C code. This should have some flexibility on what gets printed and where. The only occurance of printf(), or similar, should be in this abstraction layer. All C code should print text via macros, so they could be turned off at compile time to save resources.

What should we do with the copyright messages in the code? I suggest we just add an "and other contributors" to the end.

Should deepclone() and deepcompare() really be a part of builtin.eel?

It seems like a good idea to have the C and EEL modules work the same. The problem is that it's a bit weird to have both EEL and C code in the ./src/modules directory. Maybe we should separate them by using different terminology. We could call plain *.eel files for "libraries" and compiled C code for "modules". A library is always a plain *.eel file that is loaded run time. A module is always compiled C. If a module needs EEL code it could either pre compile it and bundle the byte code with the module (when #119 is resolved) or just install the files to a system/project wide "libraries" directory when running an "make install".

**Comment by chille** [comment]
Changed prefix from `bi_` to `loader_` in eel_loader.c

A few more things that needs to be done:
 * Separate *.pc.cmake files for each library
 * Does the EELAPI() macro need to be added to more functions?
 * Maybe have a check att `project(EEL)` and `set(PACKAGE ...)` in CMakeLists.txt

**Comment by olofson** [comment]
Yep, EELAPI() is needed for DLL exports to work on Windows! I missed some, because I used to build EEL + Eelium as one DLL, or static lib - but I've fixed some (all?) in my wip-71-rebase branch.
